File: src/frontend/cte/lib/app/data/model/cte_ferroviario_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cte/app/data/domain/domain_imports.dart';

class CteFerroviarioModel {
	int? id;
	int? idCteCabecalho;
	String? tipoTrafego;
	String? responsavelFaturamento;
	String? ferroviaEmitenteCte;
	String? fluxo;
	String? idTrem;
	double? valorFrete;

	CteFerroviarioModel({
		this.id,
		this.idCteCabecalho,
		this.tipoTrafego,
		this.responsavelFaturamento,
		this.ferroviaEmitenteCte,
		this.fluxo,
		this.idTrem,
		this.valorFrete,
	});

	static List<String> dbColumns = <String>[
		'id',
		'tipo_trafego',
		'responsavel_faturamento',
		'ferrovia_emitente_cte',
		'fluxo',
		'id_trem',
		'valor_frete',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Tipo Trafego',
		'Responsavel Faturamento',
		'Ferrovia Emitente Cte',
		'Fluxo',
		'Id Trem',
		'Valor Frete',
	];

	CteFerroviarioModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCteCabecalho = jsonData['idCteCabecalho'];
		tipoTrafego = CteFerroviarioDomain.getTipoTrafego(jsonData['tipoTrafego']);
		responsavelFaturamento = CteFerroviarioDomain.getResponsavelFaturamento(jsonData['responsavelFaturamento']);
		ferroviaEmitenteCte = CteFerroviarioDomain.getFerroviaEmitenteCte(jsonData['ferroviaEmitenteCte']);
		fluxo = jsonData['fluxo'];
		idTrem = jsonData['idTrem'];
		valorFrete = jsonData['valorFrete']?.toDouble();
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCteCabecalho'] = idCteCabecalho != 0 ? idCteCabecalho : null;
		jsonData['tipoTrafego'] = CteFerroviarioDomain.setTipoTrafego(tipoTrafego);
		jsonData['responsavelFaturamento'] = CteFerroviarioDomain.setResponsavelFaturamento(responsavelFaturamento);
		jsonData['ferroviaEmitenteCte'] = CteFerroviarioDomain.setFerroviaEmitenteCte(ferroviaEmitenteCte);
		jsonData['fluxo'] = fluxo;
		jsonData['idTrem'] = idTrem;
		jsonData['valorFrete'] = valorFrete;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCteCabecalho = plutoRow.cells['idCteCabecalho']?.value;
		tipoTrafego = plutoRow.cells['tipoTrafego']?.value != '' ? plutoRow.cells['tipoTrafego']?.value : 'AAA';
		responsavelFaturamento = plutoRow.cells['responsavelFaturamento']?.value != '' ? plutoRow.cells['responsavelFaturamento']?.value : 'AAA';
		ferroviaEmitenteCte = plutoRow.cells['ferroviaEmitenteCte']?.value != '' ? plutoRow.cells['ferroviaEmitenteCte']?.value : 'AAA';
		fluxo = plutoRow.cells['fluxo']?.value;
		idTrem = plutoRow.cells['idTrem']?.value;
		valorFrete = plutoRow.cells['valorFrete']?.value?.toDouble();
	}	

	CteFerroviarioModel clone() {
		return CteFerroviarioModel(
			id: id,
			idCteCabecalho: idCteCabecalho,
			tipoTrafego: tipoTrafego,
			responsavelFaturamento: responsavelFaturamento,
			ferroviaEmitenteCte: ferroviaEmitenteCte,
			fluxo: fluxo,
			idTrem: idTrem,
			valorFrete: valorFrete,
		);			
	}

	
}

File: backend/cte/src/model/cte_ferroviario_model.py
from src import db


class CteFerroviarioModel(db.Model):
    __tablename__ = 'cte_ferroviario'

    id = db.Column(db.Integer, primary_key=True)
    tipo_trafego = db.Column(db.String(1))
    responsavel_faturamento = db.Column(db.String(1))
    ferrovia_emitente_cte = db.Column(db.String(1))
    fluxo = db.Column(db.String(10))
    id_trem = db.Column(db.String(7))
    valor_frete = db.Column(db.Float)
    id_cte_cabecalho = db.Column(db.Integer, db.ForeignKey('cte_cabecalho.id'))


    def mapping(self, data):
        self.id = data.get('id')
        self.id_cte_cabecalho = data.get('idCteCabecalho')
        self.tipo_trafego = data.get('tipoTrafego')
        self.responsavel_faturamento = data.get('responsavelFaturamento')
        self.ferrovia_emitente_cte = data.get('ferroviaEmitenteCte')
        self.fluxo = data.get('fluxo')
        self.id_trem = data.get('idTrem')
        self.valor_frete = data.get('valorFrete')

    def serialize(self):
        return {
            'id': self.id,
            'idCteCabecalho': self.id_cte_cabecalho,
            'tipoTrafego': self.tipo_trafego,
            'responsavelFaturamento': self.responsavel_faturamento,
            'ferroviaEmitenteCte': self.ferrovia_emitente_cte,
            'fluxo': self.fluxo,
            'idTrem': self.id_trem,
            'valorFrete': self.valor_frete,
        }